### PR TITLE
Implement buffer combinator

### PR DESCRIPTION
**Pull request by NeoLegends**
This is the first combinator of #210 I'd like to have added to futures-rs. I'm kinda unsure about the name as it might be confusing in regards to `buffered` and `buffer_unordered` but since buffering is what it does, I just went with the name.

### CODE REVIEW

**Comment by NeoLegends** [comment]
Hmm, seems like sometimes the test works and sometimes it doesn't.

File: src/stream/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let maybe_next = try_ready!(self.stream.poll());
+
+        if let Some(item) = maybe_next {
+            // Push the item into the buffer and check whether it is
+            // full. If so, replace our buffer with a new and empty one
+            // and return the full one.
+            self.items.push(item);
+            if self.items.len() < self.capacity {
+                Ok(Async::NotReady)
```
Comment (tmiasko) [inline src/stream/buffer.rs]: When NotReady is returned from poll, the Stream implementation is obliged to
notify the task once a new item may be ready. As no notification of any kind is
send there, a task will wait and wait, without calling poll ever again.

What should be done here instead: the Buffer::poll should continue calling the
pull of underlying stream until NotRead is returned. Effectively offloading the
responsibility for notification to the underlying stream.

**Comment by NeoLegends** [comment]
Of course! Forgot about that.

File: src/stream/buffer.rs (hunk 1 above)
Comment (NeoLegends) [inline src/stream/buffer.rs]: Of course! Forgot about that.

**Comment by alexcrichton** [comment]
Nice! 

Perhaps this could be called `chunks` as I think it's performing similar functionality to `std::slice::chunks`? Also, could you add a test or two for the various edge cases here?

**Comment by NeoLegends** [comment]
I added an assert and a test for a capacity of zero. Which other edge cases come to your mind that are not already covered in the tests?

**Comment by alexcrichton** [comment]
Looks good to me, thanks!
